### ISSUE TITLE
Proposed Milestone/Roadmap

### ISSUE THREAD

**Issue by jbloomAus** [ai-safety-foundation/sparse_autoencoder#86]
There's a bunch of things going on here so to facilitate easy collaboration, I'm proposing a roadmap to what I see as the next big milestone. I don't think there's a huge amount of work between where we are now and this point. 

# Milestone: Gain the confidence to spend a bunch of money training SAE's on small models like Tiny Stories. 

### Why aren't we there yet?

My mental model is that no one's going to drop a bunch of cash until we know the code works and we've been trying to get aligned on how we know if it works. Below is my assessment. 

### What does reaching this milestone look like? 

I think the abstract/terminal goals here look something like:
- Loss: Reconstructed Transformer NLL is well recovered when replacing a layer with SAE output.
- Sparsity: Very few neurons are activating in the SAE in each forward pass. 
- Efficiency: The code is not highly inefficient (I suspect we’re approximately already there)
- Interpretability: We haven’t somehow managed to achieve the above criterion without achieving the valuable / currently hard-to-quantify interpretability of SAE features. 

In practice, what we want are training runs where we think we're doing well in each of those dimensions. To start, we can Reproduce [Neel’s autoencoder](https://www.lesswrong.com/posts/fKuugaxt2XLTkASkk/open-source-replication-and-commentary-on-anthropic-s#Implementation_Details) for Gelu-1L trains on 2B tokens with 16384 features. This took 12 hours on an A6000 but I think our code is likely already more efficient. 

The critical work is then to add the feedback mechanisms to be sure we are there. 

### Concrete Tasks

- SAE Training Metrics
  - [ ] [Log L0 norm of activations to wandb](https://github.com/ai-safety-foundation/sparse_autoencoder/issues/87). 
  - [ ] [Log Feature Density Histograms to wandb.](https://github.com/ai-safety-foundation/sparse_autoencoder/issues/88)
- Reconstructed Transformer Metrics
  - [ ] [Periodically run the model and compare loss with/without the SAE (measure via percent loss recovered)](https://github.com/ai-safety-foundation/sparse_autoencoder/issues/89)
Artifacts/Analysis:
  - [ ] Store SAE weights as artifacts on wandb. 
  - [ ] Add [Callum’s](https://www.perfectlynormal.co.uk/blog-sae) dashboard code to this repo and make it easy to run it on artefacts produced on any given run.  (Callum has given permission for this). 
- Benchmark Runs:
  - [ ] Reproduce [Neel’s autoencoder](https://www.lesswrong.com/posts/fKuugaxt2XLTkASkk/open-source-replication-and-commentary-on-anthropic-s#Implementation_Details) for Gelu-1L trains on 2B tokens with 16384 features.

### Feedback/Suggestions Welcome!

**Comment by lucyfarnik** [comment]
Quick suggestion, I think it would be great to think about the high-level developer interface we want this to have. I generally find it really useful to write out really neat, simple code using the library, try to make that as easy-to-understand as possible, and then just implement whatever code we need to make that happen.

What my dream demo would roughly look like:
```python
from transformer_lens import HookedTransformer
from sparse_autoencoder import SparseAutoencoder, train_sae

# Download a pre-trained SAE
sae = SparseAutoencoder.from_pretrained("gpt2-small", layer="resid_post3", expansion_ratio=8)

# Alternatively, you can train your own SAE given a specific dataset
model = HookedTransformer.from_pretrained("gpt2-small")
dataset_path = "my_dataset"
custom_sae = train_sae(model, dataset_path, layer="resid_post3", expansions_ratio=8, n_tokens=10_000_000)
```

This may be considered a separate milestone, but I think it's really good to have this high-level goal in mind when developing things

Thoughts?

**Comment by jbloomAus** [comment]
@lucyfarnik This seems fine to me and not inconsistent with the current code. I'd be interested in chunking up config into various categories too. One thing to note is that we might move away from training one SAE at a time at some point which could affect the best interface we should be wrapping with. 

@Baidicoot Thanks for the takes. I'm not sure if you're making specific recommendations about what we prioritise or do next. Eg: are you saying the most important thing is to make it so we can train SAE's on every layer at once?

A few points:
- Updating the current code to be more efficient (eg: training more than one SAE at once), seems good. Is this how your codebase currently does it? What's the trade-off with stuff like batch size? I feel this is an area where we could do lots of optimisation and I'd be excited about people doing that. 
- I should note that my back-of-envelope calculation Anthropic trained on 100B tokens for their 1L model so that would translate $600 dollars and 3 days train time. Seems like that could add up quickly over all the models we might want to do this on + different sizes of SAEs. I guess if our efficiency improves (eg: train on all the layers at once), then the cost won't be as big a deal.
- "we typically noticed that the tradeoff between sparsity and reconstruction loss was pretty strong, and doesn't seem to go away that much with more training". I might have expected more training to assist here. Are you confident this generalizes once we untie the weights and add neuron resampling? Did you use a pre-encoder bias? Regardless, are you suggesting we don't bother training for a long time? If so, what sort of impact does this have on downstream analysis?

**Comment by alan-cooney** [comment]
> Quick suggestion, I think it would be great to think about the high-level developer interface we want this to have. I generally find it really useful to write out really neat, simple code using the library, try to make that as easy-to-understand as possible, and then just implement whatever code we need to make that happen.
> 
> What my dream demo would roughly look like:
> 
> ```python
> from sparse_autoencoder import SparseAutoencoder, train_sae
> from transformer_lens import HookedTransformer
> 
> # Download a pre-trained SAE
> sae = SparseAutoencoder.from_pretrained("gpt2-small", layer="resid_post3", expansion_ratio=8)
> 
> # Alternatively, you can train your own SAE given a specific dataset
> model = HookedTransformer.from_pretrained("gpt2-small")
> dataset_path = "my_dataset"
> custom_sae = train_sae(model, dataset_path, layer="resid_post3", expansions_ratio=8, n_tokens=10_000_000)
> ```
> 
> This may be considered a separate milestone, but I think it's really good to have this high-level goal in mind when developing things
> 
> Thoughts?

I'd struggle to create the correct abstraction yet to provide a high level wrapper like this, because we don't really know how people want to experiment with SAEs yet and what direction they'll go in. This is generally why I've tried to move everything to be as flexible and modular as possible.

So for example the Loss, Optimizer and Model components are all really just PyTorch modules that can be used in any codebase. And then the other SAE-training specific components all have abstract classes so it's easy to implement a different version without digging through the rest of the codebase.

I do think we might want something like this in the future, once we have a rock solid way of training SAEs reliably across different modules. The only change I'd make at that point is then just to skip doing dynamic imports (as these break all developer tooling e.g. type checks), but your second example seems good.

**Comment by Baidicoot** [comment]
@jbloomAus ah, yeah, I should have clarified what the upshots were.

- Firstly, in 'Towards Monosemanticity: Decomposing Language Models With Dictionary Learning', Anthropic train their dictionary on 8B activations (their one-layer transformer was trained on 100B tokens, which might be where the numbers are getting mixed up). Also, if I remember correctly, I was actually training multiple autoencoders concurrently then (using this codebase), so it could be even cheaper. I will check this out.
- The main upshot I was making was: it probably won't cost all that much (I think someone threw around a $1000< figure, which I disagree with), and so we can probably get started sooner rather than later. Updating the codebase to train multiple autoencoders is as simple as making a new pipeline function, although @alan-cooney probably has better ideas on how to structure that. I'm not sure what you mean by 'trade-off with batch size'.
- What I meant re: tradeoff is that we should keep this in mind when we're working out when the training is 'good enough' (I think the current codebase is 'good enough' to start off some longer runs, since we need to check if the frontier is significantly shifted by e.g. reinit or not _when training with loads of data_.) It seems to be pretty robust to smaller changes in architecture, and centering(/pre-encoder bias). I'm not suggesting not training for a long time, since it still improves interpretability apparently, but broadly what I meant is 'there are some fundemental problems with the autoencoder approach and it has a pretty high minimum loss'.

**Comment by amirabdullah19852020** [comment]
If you just added a list of 20 models (for instance) with autoencoders you want trained for, you'd get a fair number of interested people to do it? (I've been doing a bunch of this myself the past few months with Apart Research.)

As far as the autointerp experiments go, I think they're mostly only expensive because of the GPT-4 bottleneck. If someone trains a Mistral (has someone does this already?) on autointerp, it should be much more scalable / affordable.
